### PR TITLE
Hint support burned

### PR DESCRIPTION
**Pull request by amrondonp**
The first commit are the files that the people from codemirror updated. The second commit contains the files for making autocomplete available and the third contains the front-end changes in inginious.

@0xDCA

### CODE REVIEW

**Comment by 0xDCA** [comment]
I am a bit confused by this. Where did you add the autocomplete hints? I see that you added inginious/frontend/common/static/js/codemirror/addon/hint/generic-hint.js . Is this yours or does it come from the CodeMirror codebase? If it's the former, you should keep it separate from the CodeMirror code

**Comment by amrondonp** [comment]
generic complete is a piece of code that I wrote entirely. However, as it's being used and referenced in show-hint.js (principal autocomplete file) so that every language has the generic autocomplete support. I think we should leave it under addons

File: inginious/frontend/common/static/js/extensions/codemirror/generic-hint.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  else if (typeof define == "function" && define.amd) // AMD
+    define(["../../lib/codemirror"], mod);
+  else // Plain browser env
+    mod(CodeMirror);
+})(function (CodeMirror) {
+  var Pos = CodeMirror.Pos;
+
+  function shouldIgnoreToken(token) {
+    return token.type === null
+           || token.type == "string"
```
Comment (0xDCA) [inline inginious/frontend/common/static/js/extensions/codemirror/generic-hint.js:16]: Are these token types consistent across languages?

File: inginious/frontend/common/static/js/extensions/codemirror/generic-hint.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    return {
+      list: selectedWords,
+      from: Pos(cursor.line, editor.getTokenAt(cursor).start),
+      to: Pos(cursor.line, editor.getTokenAt(cursor).end)
+    };
+  }
+  
+  CodeMirror.registerHelper("hint", "python", genericHint);
+  CodeMirror.registerHelper("hint", "clike", genericHint);
+  CodeMirror.registerHelper("hint", "ruby", genericHint);
```
Comment (0xDCA) [inline inginious/frontend/common/static/js/extensions/codemirror/generic-hint.js:82]: Is Java missing here?

File: inginious/frontend/common/static/js/extensions/codemirror/generic-hint.js (hunk 1 above)
Comment (0xDCA) [inline inginious/frontend/common/static/js/extensions/codemirror/generic-hint.js]: these*

Comment (amrondonp) [reply inginious/frontend/common/static/js/extensions/codemirror/generic-hint.js:16]: Yes, It has been tested in the languages that we are interested in

File: inginious/frontend/common/static/js/extensions/codemirror/generic-hint.js (hunk 2 above)
Comment (amrondonp) [reply inginious/frontend/common/static/js/extensions/codemirror/generic-hint.js:82]: clike is a "metamode" that references C#, Java, C, Cpp

**Comment by 0xDCA** [comment]
Cool, LGTM

**Pull request merged**